### ISSUE TITLE
aarch64: Multiprocessing on a single CPU can cause the parent process to die.

### ISSUE THREAD

**Issue by chl337** [QuarkContainer/Quark#1210]
When running on one (v)CPU, the chain:

a)P: (user-space)run
b)P: clone
c)P: wait...
d)C: run
e)C: exit
f)P: wait - resume
g)P: (user-space) run

breaks on step g) as parent returns to user-space. Most usual error, parent tries to access a non valid address (suspiciously the value is =0x2, which is the same as the PID of the child in). I have checked the correctness of _PtRegs_ before and after the _wait()_ syscall and the for changes on the user-space stack. The _PtRegs_ look correct and the stack gets COW.

[quark.log](https://github.com/QuarkContainer/Quark/files/15192554/quark.log)

[test program](https://www.online-cpp.com/EVWsjt2ymu)

I would appreciate any help @CharlyYu, @QuarkContainer.

**Comment by shrik3** [comment]
test program is "deleted"

**Comment by shrik3** [comment]
I worked out a fix. I'll try to submit the changes asap.

I have two tests. Both run to completion without issue.

test_pthread.c
```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>
#include <time.h>
#include <sys/syscall.h>

// The function to be executed by all threads
void *myThreadFun(void *vargp)
{
	pid_t pid = getpid();
	pid_t tid = syscall(SYS_gettid);
	gid_t gid = getgid();
	int i;
	// Let us create a static variable to observe its changes
	static int s = 0;

	// Change static and global variables

	void *sp;
	__asm__ volatile ("mov %0, sp\n" : "=r" (sp) );
	// Print the argument, static and global variables
	for(i=1;i<3;i++){
		++s; ++g;
		printf("PID: %d, TID: %d, GID: %d, SP: %p, Static: %d, Global: %d\n", pid, tid, gid, sp ,++s, ++g);
		sleep(random()%3);
	}
}

int main()
{
	printf("testing pthread\n");
	int i;
	pthread_t threads[6];
	// Let us create three threads
	for (i = 0; i < 6; i++){
		pthread_create(&threads[i], NULL, myThreadFun, (void *)&threads[i]);
	}
	for (i = 0; i < 6; i++){
		pthread_join(threads[i], NULL);
	}
	return 0;
}
```

test_wait.c
```c
#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <string.h>
#include <sys/wait.h>
#include <unistd.h>
#include <time.h>

#define btoa(x) ((x)?"true":"false")

void wstatus_info(int ws){
	if(WIFEXITED(ws)){
		printf("child exited %d\n", WEXITSTATUS(ws));
	}
	if(WIFSIGNALED(ws)){
		int sig = WTERMSIG(ws);
		printf("child term-signaled %d %s\n", sig, strsignal(sig));
	}
}

int main() {
	pid_t cpids[5];
	int status = 0;
	int i;

	// void *sp;
	// __asm__ volatile ("mov %0, sp\n" : "=r" (sp) );
	for (i=0; i<5; i++){
		pid_t p = fork();
		if (p<0){
			printf("fork failed!");
			exit(1);
		}
		if (p==0){
			int j;
			p = getpid();
			for (j=0;j<3;j++){
				printf("[CHILD] I AM %d\n", p);
				sleep(1);
			}
			exit(0);
		} else {
			cpids[i] = p;
			printf("[PAREN] forked child with pid %d\n", p);
		}
	}

	for (i=0; i<5; i++){
		pid_t pw = waitpid(cpids[i], &status, 0);
		printf("[PAREN]. waitpid(%d) returned %d, status=%d\n", cpids[i], pw, status);
		if (status !=0){
			wstatus_info(status);
		}
	}
	// should not reach.
	return 0;
}
```

**Comment by chl337** [comment]
Then I will close the issue after you submit the changes.

**Comment by chl337** [comment]
I tested the 'offline' fixes locally, they work for me too.

**Comment by shrik3** [comment]
> Then I will close the issue after you submit the changes.

I think the remaining issue is on signaling. You are already working on that I suppose?

**Comment by chl337** [comment]
> > Then I will close the issue after you submit the changes.
> 
> I think the remaining issue is on signaling. You are already working on that I suppose?

I have opened #1211 to track issues regarding signals.

**Comment by chl337** [comment]
Closed from #1212
